Support centralized configuration in operator

nicolaferraro opened the pull request:
## Cover letter

This adds support for centralized configuration in the operator. Users can change the CR to update the cluster configuration and have it synchronized to the cluster, without triggering a restart if not needed.

The central piece introduced here is an object called GlobalConfiguration which the operator uses to read/write both node/local and clusterwide configuration properties. The GlobalConfiguration has different strategies (modes) to deal with old (pre v22) clusters, where all properties go to `redpanda.yaml`, new clusters where configuration is split between `redpanda.yaml` and `.bootstrap.yaml` and contains also a transitioning (mixed, default) strategy that uses both files, to simplify rolling upgrades from older versions.

The management of the configuration aspect is delegated to a pseudo-controller, meaning that it's implemented like a secondary controller, but its code is executed after the main controller. The reason is that we probably need to switch to server side apply and start doing patches to resources instead of full rewrites before we can introduce a real secondary controller, otherwise there'll be too many concurrent modification conflicts.
To manage particular fields of the resources, the main reconcile loop is instructed to not touch some fields when reconciling them, while the configuration controller modify them directly (mimics what server side apply does naturally).

There are now two digest computed by the controller, one for node properties, one for central properties requiring restart. When those change, a restart of the cluster is automatically triggered. Centralized properties that don't require a restart are just updated via the admin API.

Patches sent to the admin API are computed using a three (or four) way merge, so that only properties changed in the CR are changed by the operator, leaving other workflows (rpk, kafka api) that can change configuration unaffected. The last-applied-configuration is tracked in the configmap for this reason.

The state of the cluster configuration is fully mapped in a condition.

I've found many issues with serialization and tried to fix them. The main problem is that most of the centralized configuration is now unstructured and when e.g. retrieving the config from the admin API the Go json deserialization defaults to using `float64` for all numeric values in unknown properties. ~I've switched to use json.Number when possible to avoid losing information, but still there are issues with yaml serialization that does not understand the type.~ I needed to address these issues because both detection of changes in configuration and also computation of patches should not be affected by the way we represent numbers. More tests and refactoring are needed for these aspects.

I've included a long `envtest` test for checking the behavior, plus some `kuttl` e2e tests.


Main remaining things:

- [x] Add more tests for floating point properties
- [x] Support admin API TLS and basic authentication (maybe in another PR)

## Release notes

* Added support for centralized configuration in the operator

0x5d reviewed (changes requested):
This is a great change, thanks!
On a related topic, this is a _great_ (very big) PR 😄 The first commit took me very long to get through, so I couldn't review the whole thing in the time I allocated for it. In the future please consider splitting changes in smaller commits, so that it's easier to focus on specific areas. Github doesn't really help either, since it just sorts the files alphabetically, leaving it up to the reader to find the place to start.

jcsp commented on src/go/k8s/pkg/resources/configuration/configuration.go:
```diff
@@ -0,0 +1,154 @@
+
+// otherKnownNodeProperties contains node properties that are not listed in the struct
+var otherKnownNodeProperties = map[string]bool{
+	"redpanda.enable_central_config": true, // deprecated but may be needed in some tests
```
This is no longer needed with latest redpanda, all references to `enable_central_config` can be removed.

jcsp reviewed (commented):
I didn't get very far through this yet (67 files!) but publishing a couple of comments rather than leaving them hidden in a draft review

jcsp commented on src/go/k8s/controllers/redpanda/cluster_controller_configuration.go:
```diff
@@ -28,34 +28,35 @@ func (r *ClusterReconciler) markConfigurationChanged(
+		} else if config != nil {
+			if config.ClusterConfiguration == nil {
+				// Upgrade scenario: current cluster is not using centralized configuration.
+				// We need to extract bootstrap properties from the old format
```
Bootstrap is only needed on first startup of a cluster: if a cluster is being upgraded from an earlier version, then you can skip it entirely (it's fine for the file to simply not exist).

nicolaferraro replied:
Right, the comment was not precise. The state indicates a specific upgrade from a version that was not using centralized configuration e.g. 21.11.1 to 22.1.1, so I think the bootstrap file will be used in this case by the new version.

I preferred anyway to always write the bootstrap file in the configmap, both to track the configuration more easily, but also to detect configuration drifts without having to call the admin API (which may not be available during some reconciliation loops).

I've changed this computation a bit in a [subsequent commit](https://github.com/redpanda-data/redpanda/pull/3978/commits/28a6ed92c761345bbf48f903a465152355411ecf) to simplify it.

nicolaferraro replied:
I should definitely change the term "drift", also in the code. The operator is not checking if the cluster is still configured as expected (as the term "drift" may imply), it's checking if the expected cluster configuration is different from the last-submitted one (stored in the configmap). Only after a "change" has been detected, the admin API is checked and any drift is fixed.

But e.g. if I set on the CR additional config `redpanda.segment_appender_flush_timeout_ms=2000` and someone changes that value manually to `3000` in the admin API, the operator will not set it back to `2000` until there's another change in the CR config section that triggers the merge process. We said that the users must be aware that they should not change the same configuration property in multiple places.

I did this verification on the CM instead of the cluster so that we're sure that, even in corner cases, the reconcile loop is not affected by the cluster state, i.e. it continues to work correctly when cluster is starting/restarting.

If detecting real drifts against the cluster is a desired feature, it's still possible to add a periodic check. Wdyt @RafalKorepta ?

nicolaferraro commented on src/go/k8s/config/manager/kustomization.yaml:
```diff
@@ -13,4 +13,4 @@ kind: Kustomization
 images:
 - name: vectorized/redpanda-operator
   newName: vectorized/redpanda-operator
-  newTag: latest
```
Uh, thanks for noticing. This change was not supposed to be committed.

nicolaferraro commented on src/go/k8s/controllers/redpanda/cluster_controller.go:
```diff
@@ -44,10 +45,11 @@ var (
+	configuratorSettings  resources.ConfiguratorSettings
+	clusterDomain         string
+	Scheme                *runtime.Scheme
+	AdminAPIClientFactory utils.AdminAPIClientFactory
```
Good question.. In general it's because the controller is singleton, while the admin API client depends on the cluster to contact. Anyway, I could have put the client directly without using the factory, since I inject a different one only in tests, but it didn't make a lot of sense outside the testing scope..

0x5d replied:
I thought it was a little confusing that it's checked if `AdminAPIClientFactory` is `nil`, and if it is, then NewAdminAPI is called directly. It's an unexpected side effect in my opinion.

nicolaferraro replied:
Yeah, agreed. Better setting it always.

nicolaferraro commented on src/go/k8s/controllers/redpanda/cluster_controller_configuration.go:
```diff
@@ -0,0 +1,275 @@
+		if err := statefulSetResource.SetCentralizedConfigurationHashInCluster(ctx, hash); err != nil {
+			return fmt.Errorf("could not update config hash on statefulset %s/%s: %w", redpandaCluster.Namespace, redpandaCluster.Name, err)
+		}
+	}
```
I think this has been addressed in the core with validation. Once the cluster is started for the very first time, the only way to change the centralized configuration is via admin API (.bootstrap.yaml file has no effect), so the cluster will never accept a configuration that could lead to a bootstrap crash.. or at least, it would be a bug to fix immediately..

jcsp commented on src/go/rpk/pkg/cli/cmd/redpanda/mode_test.go:
```diff
@@ -29,7 +29,7 @@ import (
 func fillRpkConfig(path, mode string) *config.Config {
 	conf := config.Default()
 	val := mode == config.ModeProd
-	conf.Redpanda.DeveloperMode = !val
```
`developer_mode` ended up as a node config property, so you can undo the part of this that removed it from the node config object

jcsp commented on src/go/k8s/pkg/resources/configmap.go:
```diff
@@ -681,14 +711,39 @@ func (r *ConfigMapResource) GetLastAppliedConfigurationFromCluster(
 	return nil, false, nil
 }
 
+// cleanupConfig translates json Number objects into int64, otherwise yaml.v3
```
I'm not sure of the context where this is used, but in general it isn't correct to assume all numbers are integers: floating point configuration properties do exist.  The config schema distinguishes between `number` and `integer` types.

nicolaferraro replied:
Yes, I've noticed the presence of some floating point properties and fixed this later.

jcsp commented on src/go/k8s/pkg/resources/configuration/patch.go:
```diff
@@ -68,10 +70,39 @@ func FourWayMerge(
+// - Integers and their related string representations should be considered equal
+// - Floating point values should match their integer or string representations if close to an integer number
+// - JSON Number objects (e.g. obtained from JSON APIs) should match their string, integer or float representations
+func looseEqual(v1, v2 interface{}) bool {
```
The schema defines a type for each property, so you could normalize all the values via that type knowledge rather than doing a totally general comparison.  e.g. if a field is of type `number` then you can trust the value that the config API gives you will be a floating point, and parse your user input into a floating point before doing the comparison.

nicolaferraro replied:
This is a good point. For total correctness, I may need to change the rpk config interface to deserialize numbers to json.Number (a.k.a. strings), otherwise it will translate integers to float64. I did it in the draft but I reverted it because of multiple failing tests (in core).

jcsp commented on src/go/k8s/controllers/redpanda/cluster_controller_configuration.go:
```diff
@@ -0,0 +1,275 @@
+	if !reflect.DeepEqual(originalCondition, currentCondition) || (stsNeedsRestart && !redpandaCluster.Status.Upgrading) {
+		// Trigger restart here if needed
+		if stsNeedsRestart {
+			redpandaCluster.Status.Upgrading = true
```
Is the assumption here that rolling restarts are going to be totally safe and transparent by the time this is released?  Otherwise it's a bit risky to auto-restart on config edits.

Also, it seems like the "Upgrading" field should be renamed if it is also used to trigger restarts generally

nicolaferraro replied:
Afaik auto-restart is the default behavior even without these changes. Is this correct @RafalKorepta ?

The upgrading field is misleading, right. Something like `restarting` (if a field is needed) would be better. I started to work on rolling restarts so I may fix the field in that PR.

jcsp commented:
The release nodes need some expansion to really describe the behavioral changes the user will see: e.g. cluster restarts on config edits are a big user-visible change.

jcsp commented:
I've done a pass through the code and commented where things jumped out at me, but for correctness this needs a review from someone more familiar with the operator (@alenkacz or @RafalKorepta perhaps?)

nicolaferraro commented on src/go/k8s/controllers/redpanda/cluster_controller_configuration.go:
```diff
@@ -0,0 +1,275 @@
+	}
+
+	if oldHash != hash {
+		// Definitely needs restart
```
The hash for centralized configuration is only computed with fields with `needs_restart=true` in the schema, so restarts will be triggered only when one of that fields change. There's a specific test for this: https://github.com/redpanda-data/redpanda/pull/3978/files#diff-0cfeba1779bdecfec8476b8602a73be2ab7de5892aec5463b2866bc66f4b42b2R222

RafalKorepta commented on src/go/k8s/controllers/redpanda/cluster_controller_configuration.go:
```diff
@@ -0,0 +1,275 @@
+// preCheckConfigurationChange verifies and marks the cluster as needing synchronization (using the ClusterConfigured condition).
+// When this method returns true, the cluster CR is marked, indicating that the target cluster will be eventually synchronized (via reconcileConfiguration).
+func (r *ClusterReconciler) markConfigurationChanged(
+	ctx context.Context,
```
I had hard time to understand in what condition the annotation is applied to the config map. Could you elaborate in this godoc why and when the config map annotation plays the role into reconciliation?

From current description I would assume only Cluster CR condition is updated, but there is fancy way to update annotation too.

nicolaferraro commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_types_test.go:
```diff
@@ -131,3 +133,41 @@ func TestRedpandaResourceRequirements(t *testing.T) {
+	assert.Equal(t, "reason", cond.Reason)
+	assert.Equal(t, "message", cond.Message)
+	condTime := cond.LastTransitionTime
+	assert.True(t, condTime.Add(-1*time.Millisecond).Before(time.Now()))
```
I guess I could use `!time.After` :smile:

nicolaferraro commented on src/go/k8s/controllers/redpanda/cluster_controller.go:
```diff
@@ -268,6 +276,19 @@ func (r *ClusterReconciler) Reconcile(
+	if errors.As(err, &requeueErr) {
+		log.Info(requeueErr.Error())
+		return ctrl.Result{RequeueAfter: requeueErr.RequeueAfter}, nil
+	}
```
Good finding. Thanks!

nicolaferraro commented on src/go/k8s/pkg/resources/configmap.go:
```diff
@@ -631,15 +617,131 @@ func generatePassword(length int) (string, error) {
+		return fmt.Errorf("could not marhsal configuration: %w", err)
+	}
+	newAnnotation := string(ser)
+	if existing.Annotations[LastAppliedConfigurationAnnotationKey] != newAnnotation {
+		if existing.Annotations == nil {
```
Mmh, no a nil map behaves like an empty map, until you try to *write* to it (in which case it panics)

RafalKorepta replied:
THX :)

nicolaferraro commented on src/go/k8s/kuttl-test.yaml:
```diff
@@ -15,6 +15,8 @@ commands:
   - command: "make deploy"
   - command: "./hack/wait-for-webhook-ready.sh"
   - command: "mkdir -p tests/_e2e_artifacts"
+  - command: "docker tag localhost/redpanda:dev vectorized/redpanda:v22.1.1-dev"
```
I think I can get rid of it. The reason is that there's a feature gates that is enabled only for v22+ clusters, but I can extend it also to the `dev` version and avoid creating the alias..

RafalKorepta commented on src/go/k8s/pkg/resources/configuration/patch_test.go:
```diff
@@ -148,6 +149,18 @@ func TestLooseEquality(t *testing.T) {
 			v1: json.Number("536870912"),
 			v2: 536870912,
 		},
+		{
+			v1: json.Number("0.9999"),
+			v2: 0.9999,
+		},
+		{
+			v1: json.Number("0.9999"),
+			v2: 0.9999,
+		},
```
I don't see the difference between those 2 test cases.

RafalKorepta reviewed (changes requested):
I think the changes are good. Please rewrite the git history, so that ppl will have easier time to dissect and do `git bisect`.

nicolaferraro commented on src/go/rpk/pkg/config/schema.go:
```diff
@@ -43,31 +43,24 @@ func (c *Config) File() *Config {
-	AdvertisedKafkaApi                   []NamedSocketAddress   `yaml:"advertised_kafka_api,omitempty" mapstructure:"advertised_kafka_api,omitempty" json:"advertisedKafkaApi,omitempty"`
-
-	Other                                map[string]interface{} `yaml:",inline" mapstructure:",remain"`
+	Directory                  string                 `yaml:"data_directory" mapstructure:"data_directory" json:"dataDirectory"`
+	Id                         int                    `yaml:"node_id" mapstructure:"node_id" json:"id"`
+	Rack                       string                 `yaml:"rack,omitempty" mapstructure:"rack" json:"rack"`
+	SeedServers                []SeedServer           `yaml:"seed_servers" mapstructure:"seed_servers" json:"seedServers"`
+	RPCServer                  SocketAddress          `yaml:"rpc_server" mapstructure:"rpc_server" json:"rpcServer"`
+	RPCServerTLS               []ServerTLS            `yaml:"rpc_server_tls,omitempty" mapstructure:"rpc_server_tls,omitempty" json:"rpcServerTls"`
+	KafkaApi                   []NamedSocketAddress   `yaml:"kafka_api" mapstructure:"kafka_api" json:"kafkaApi"`
+	KafkaApiTLS                []ServerTLS            `yaml:"kafka_api_tls,omitempty" mapstructure:"kafka_api_tls,omitempty" json:"kafkaApiTls"`
+	AdminApi                   []NamedSocketAddress   `yaml:"admin" mapstructure:"admin" json:"admin"`
+	AdminApiTLS                []ServerTLS            `yaml:"admin_api_tls,omitempty" mapstructure:"admin_api_tls,omitempty" json:"adminApiTls"`
+	CoprocSupervisorServer     SocketAddress          `yaml:"coproc_supervisor_server,omitempty" mapstructure:"coproc_supervisor_server" json:"coprocSupervisorServer"`
+	AdminAPIDocDir             string                 `yaml:"admin_api_doc_dir,omitempty" mapstructure:"admin_api_doc_dir" json:"adminApiDocDir"`
+	DashboardDir               string                 `yaml:"dashboard_dir,omitempty" mapstructure:"dashboard_dir" json:"dashboardDir"`
+	CloudStorageCacheDirectory string                 `yaml:"cloud_storage_cache_directory,omitempty" mapstructure:"cloud_storage_cache_directory" json:"CloudStorageCacheDirectory"`
+	AdvertisedRPCAPI           *SocketAddress         `yaml:"advertised_rpc_api,omitempty" mapstructure:"advertised_rpc_api,omitempty" json:"advertisedRpcApi,omitempty"`
+	AdvertisedKafkaApi         []NamedSocketAddress   `yaml:"advertised_kafka_api,omitempty" mapstructure:"advertised_kafka_api,omitempty" json:"advertisedKafkaApi,omitempty"`
+	DeveloperMode              bool                   `yaml:"developer_mode" mapstructure:"developer_mode" json:"developerMode"`
+	Other                      map[string]interface{} `yaml:",inline" mapstructure:",remain"`
```
There was a formatting issue, `gofmt` recomputes spaces from scratch after an empty line, so everything became unreadable..

nicolaferraro commented on src/go/rpk/pkg/config/params.go:
```diff
@@ -248,7 +248,7 @@ func (p *Params) Load(fs afero.Fs) (*Config, error) {
 				Port:    9644,
 			}}},
-			Other: map[string]interface{}{"developer_mode": true},
+			DeveloperMode: true,
```
Developer mode became later a node property again, I'll try to merge the two commits

nicolaferraro commented on src/go/k8s/controllers/redpanda/cluster_controller_configuration.go:
```diff
@@ -110,75 +98,174 @@ func (r *ClusterReconciler) reconcileConfiguration(
+		return err
+	}
+
+	// TODO a failure and restart here (after successful patch, before setting the last applied configuration) may lead to inconsistency if the user
+	// changes the CR in the meantime (e.g. removing a field), since we applied a config to the cluster but did not store the information anywhere else.
+	// A possible fix is doing a two-phase commit (first stage commit on configmap, then apply it to the cluster, with possibility to recover on failure),
+	// but it seems overkill given that the case is rare and requires cooperation from the user.
```
Well, I was thinking to an edge case like this:

- User starts a new cluster and configuration is synchronized
- User sets `redpanda.myproperty=value`
- The reconciler computes a patch and sends it to the cluster, success
- The reconciler tries to store in the CM the last configuration saved in the cluster, but it fails
- Before the reconciliaiton loop restarts, the user changes his mind and deletes the change to `myproperty` (it works if he just modify the value)

Now, when the reconciler analyzes the configuration again, it notices the change, but when comparing current config with the one on the cluster, it thinks that `myproperty` has been changed by someone else (there's no track of it) and leaves it as is.

The possible solution I was hinting at was..
- Track that I'm going to write `redpanda.myproperty=value`
- Write `redpanda.myproperty=value` to the cluster
- Store last configuration applied
- If something goes wrong, recover

But it does not seem a realistic problem..

nicolaferraro commented:
I should have addressed all the comments. I also tried to rewrite the history to be easier to understand (for what is possible).

When I enabled the centralized configuration feature also on the `dev` tag, the CI found some regressions that I had to fix changing some logic a bit.
E.g. the pre-check (to detect configuration changes) that was done before the main reconcile loop changed the resources, has been moved inside the configmap reconciliation, because in some cases the computation of the current configuration relies on secrets being previously stored in the namespace.

It would be great if you have a second look: @0x5d, @RafalKorepta, @jcsp

RafalKorepta reviewed (commented):
It looks good, left few comments

nicolaferraro commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_types.go:
```diff
@@ -285,8 +286,111 @@ type ClusterStatus struct {
+}
+
+// ClusterConditionType is a valid value for ClusterCondition.Type
+// +enum
```
Indeed, I took inspiration from kube structs when creating the condition, but the `+enum` tag is only known by [kube-openapi](https://github.com/kubernetes/kube-openapi/blob/94424faaec1fca6f0173d12c988a642fcc0806d8/test/integration/testdata/enumtype/enum.go). I'll replace it with the Kubebuilder equivalent.

nicolaferraro commented on src/go/k8s/pkg/resources/configuration/configuration.go:
```diff
@@ -0,0 +1,146 @@
+	if v, ok := knownNodeProperties[prop]; ok {
+		return v
+	}
+	for k := range knownNodeProperties {
+		if strings.HasPrefix(prop, fmt.Sprintf("%s.", k)) {
+			return true
+		}
+	}
```
There are some examples in the tests:

```
	"redpanda.cloud_storage_cache_directory": "/tmp",
	"redpanda.rpc_server.port":               "8080",
```

Everything under `redpanda.` is a cluster property, except for those explicitly declared in the struct. There can also be complex object, like in the case of `rpc_server`.

nicolaferraro replied:
While e.g.

```
"redpanda.cloud_storage_region":          "us-west-1",
```

Sets a cluster property.

nicolaferraro commented on src/go/k8s/pkg/resources/configuration/configuration_modes.go:
```diff
@@ -0,0 +1,192 @@
+	GlobalConfigurationModeClassic GlobalConfigurationMode = globalConfigurationModeClassic{}
+	// GlobalConfigurationModeCentralized splits properties between "redpanda.yaml" and ".bootstrap.yaml"
+	GlobalConfigurationModeCentralized GlobalConfigurationMode = globalConfigurationModeCentralized{}
+	// GlobalConfigurationModeMixed puts central configuration properties in both "redpanda.yaml" and ".bootstrap.yaml"
```
Node/local configuration goes in `redpanda.yaml`, and it can contain also central configuration properties. While `.bootstrap.yaml` can only contain central configuration properties.

The format is also different. E.g.

`redpanda.yaml` has central properties nested under `redpanda`:

```yaml
# ...
redpanda:
  cloud_storage_region: us-west-1
```

While `.bootstrap.yaml` is k/v:

```
cloud_storage_region: us-west-1
```

Properties like `redpanda.cloud_storage_cache_directory` can only go in `redpanda.yaml`.

nicolaferraro commented on src/go/k8s/pkg/resources/configmap.go:
```diff
@@ -121,18 +126,73 @@ func (r *ConfigMapResource) Ensure(ctx context.Context) error {
+	c k8sclient.Client,
+	logger logr.Logger,
+) error {
+	// Do not touch existing last-applied-configuration (it's not reconciled in the main loop)
+	if val, ok := current.Annotations[LastAppliedConfigurationAnnotationKey]; ok {
+		if modified.Annotations == nil {
+			modified.Annotations = make(map[string]string)
+		}
+		modified.Annotations[LastAppliedConfigurationAnnotationKey] = val
+	}
```
No, probably you missed that the target is different: `current` vs `modified`. The meaning is that if the configmap in the cluster contains a value for that annotation, then the configmap we're going to apply will have the same value (since we use .Update instead of .Patch, we need to provide a full configmap when we apply it).

nicolaferraro commented on src/go/k8s/pkg/resources/configmap.go:
```diff
@@ -266,34 +333,39 @@ func (r *ConfigMapResource) createConfiguration(
 
-	for _, user := range r.pandaCluster.Spec.Superusers {
-		cr.Superusers = append(cr.Superusers, user.Username)
+	if len(r.pandaCluster.Spec.Superusers) > 0 {
+		superusers, err := getSuperusersFromConfig(cfg)
+		if err != nil {
+			return nil, err
+		}
+		for _, user := range r.pandaCluster.Spec.Superusers {
+			superusers = append(superusers, user.Username)
+		}
```
So there are 3 places where superusers are computed. The CR is one of them, but when you enable pandaproxy or schema registry, their users are also added to the list of superusers. So the code you see here is the equivalent of an `append` on an unstructured property.
It's true that in this particular spot the slice is supposed to be empty, but doing an append makes the code more robust to refactorings imo.

Btw, I think it's a good idea to put an `append` method for string slices in the global property object, so that I can avoid the boilerplate and also test it...

RafalKorepta replied:
Thank you for explanation

RafalKorepta reviewed (approved):
LGTM

RafalKorepta reviewed (changes requested):
Sorry I just realised that some new files doesn't have copyrights.

Can you follow what was done in this PR https://github.com/redpanda-data/redpanda/pull/4056


```
/*
 * Copyright 2021 Vectorized, Inc.
 *
 * Licensed as a Redpanda Enterprise file under the Redpanda Community
 * License (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
 */
```

RafalKorepta reviewed (approved):
The copyrights will be addressed in the next follow up PR.

cc @andrewhsu
